pagopa/io-react-native-login-utils
# [IOBP-524] Allow user to switch apps while authenticating on Android

## Short description
This PR fixes an issue on Android devices where users were unable to switch apps during the authentication session. With this fix the browser is able to remain active while the user navigates to other apps.

The solution takes inspiration from a similar issue encountered by AppAuth for Android library ([here the PR](https://github.com/openid/AppAuth-Android/pull/109)). It uses an intermediary Activity, `AuthenticationManagerActivity`, to allow the Chrome Custom Tab to be started without the `NEW_ACTIVITY_NO_HISTORY` flag, and the `RedirectUriReceiverActivity`, which allows to remove the Browser activity from the recents.

![Untitled-2024-01-31-1157](https://github.com/pagopa/io-react-native-login-utils/assets/6160324/d5c87d87-94db-4954-b144-6747673fb182)

<details><summary>Mermaid Flow</summary>
<p>

```mermaid
sequenceDiagram
    participant ReactAppMainActivity
    participant openAuthenticationSession as openAuthenticationSessio(io-react-native-login-utils)
    participant AuthorizationService as AuthorizationService(io-react-native-login-utils)
    participant BrowserHandler as BrowserHandler(io-react-native-login-utils)
    participant CustomTabsServiceConnection as CustomTabsServiceConnection(io-react-native-login-utils)
    participant AuthManagerActivity as AuthManagerActivity(io-react-native-login-utils)
    participant ChromeCustomTabs
    participant RedirectUriActivity as RedirectUriActivity(io-react-native-login-utils)

    ReactAppMainActivity->>openAuthenticationSession: Calls with auth URL
    openAuthenticationSession->>AuthorizationService: Instantiates
    AuthorizationService->>BrowserHandler: Instantiates
    BrowserHandler->>CustomTabsServiceConnection: Instantiates
    BrowserHandler->>ReactAppMainActivity: Do bind service
    ReactAppMainActivity->>CustomTabsServiceConnection: Bind    
    AuthorizationService->>AuthManagerActivity: performAuthRequest with authIntent and start    
    AuthorizationService->>AuthorizationService: Disposing...
    AuthorizationService->>BrowserHandler: Unbind        
    BrowserHandler->>ReactAppMainActivity: Do unbind service
    ReactAppMainActivity->>CustomTabsServiceConnection: Unbind
    AuthorizationService->>AuthorizationService: Disposed
    AuthManagerActivity->>ChromeCustomTabs: Starts authIntent
    Note right of AuthManagerActivity: Saves state of authIntent and tracks if auth request has previously started.
    ChromeCustomTabs->>RedirectUriActivity: The auth flow ends with an app link that trigger
    RedirectUriActivity->>AuthManagerActivity: Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP
    RedirectUriActivity->>RedirectUriActivity: Kills itself    
    AuthManagerActivity->>AuthManagerActivity: Handles response
    Note over AuthManagerActivity: Restore state of authIntent and if an auth request has previously started.
    AuthManagerActivity->>ReactAppMainActivity: resolve/reject promise
    AuthManagerActivity->>AuthManagerActivity: kills itself thus removing ChromeCustomTabs from history

    Note over ReactAppMainActivity, RedirectUriActivity: Flow initiates in ReactAppMainActivity and transitions through library components to manage authentication via ChromeCustomTabs. RedirectUriActivity handles the final redirect URI to adjust the navigation stack

```

</p>
</details> 


## List of changes proposed in this pull request
- Added `AuthorizationManagerActivity` which is responsible to prepare and launch the auth Intent (`CustomTabsIntent`)
- Added `RedirectUriReceiverActivity` which handles the result from the `CustomTabsIntent` via deeplink
- Improved error management

## How to test
On **Android**, within the example app, start the [io-dev-api-server](https://github.com/pagopa/io-dev-api-server) and try to login and/or perform a payment request. You should be able to navigate to other apps while the Chrome custom tab is open and  you should be able to complete the authentication/authorization when returning back to the example app.

## Preview

<video src="https://github.com/pagopa/io-react-native-login-utils/assets/6160324/c435df3c-4939-4e6b-b1bb-212b0b853f4f" />
